scripts/mkstatic.py

```diff
@@ -896,7 +896,7 @@ class Group(object):
    private
    public :: {subroutines}
 
-   logical, save :: initialized = .false.
+   logical, dimension(200), save :: initialized = .false.
```

Comment:
The `save` is redundant, it is on by default at module scope.
How / where is the index initialized (there is nothing in this PR to indicate that as far as I can tell)?
Could the `200` limit be a public parameter that can be checked somewhere in case of an overflow?
Better yet, is it known when the `write` method is called? The actual size could be passed in.

Reply:
I agree that having this as a public parameter is better than hard-coding it to 200 (though that's plenty high enough for NRL's purposes).  I'm at the edge of my practical knowledge of CCPP's innards and would need to spend some time figuring out how to add the size parameter.  Someone else reading this would probably know how to do it in 10 minutes.  If so, I support that.

Reply:
@michalakes I have opened a PR onto your branch that makes this value a little less hard-coded (defined in common.py). Ideally it would have been defined in the host model's config file, but it looks like the order of operations in the ccpp_prebuild.py script doesn't allow for that without relatively invasive changes, and that doesn't seem worth the effort for the prebuild script. 

@gold2718 @climbfuji Let me know if you have any comments on this, or if there's something I'm missing. It does seem like it should be relatively easy to define this variable in the host config file with capgen, since all the config variables are available in the `run_env` structure.

Reply:
@mkavulich,

> The `save` is redundant, it is on by default at module scope. How / where is the index initialized (there is nothing in this PR to indicate that as far as I can tell)? Could the `200` limit be a public parameter that can be checked somewhere in case of an overflow? Better yet, is it known when the `write` method is called? The actual size could be passed in.

While the size can easily be entered and entered around passed around with capgen, that will not help with either non-compliant parameterizations or with capgen itself.
capgen makes use of module variables to handle automatic memory management for variables that need to be passed between suite phases (e.g., created at init time and used at run time). To accomodate multiple instances would require a pretty extensive reworking of this infrastructure.

Reply:
@michalakes ,

Trying a different approach:
Let's say there is a 3-D variable that is produced by an init routine (`intent(out)`) and used by a run routine (e.g., `intent(inout)`). This module variable (in capgen) is initialized by every instance.  _Every_ instance would (re-)initialize the _same_ variable. At run time, every instance would again use and modify the same memory location. How does creating a logical array address this situation?

Reply:
@gold2718, I don't follow your example.  My understanding of the purpose of the logical is to indicate if shared data that gets initialized/ingested only occurs once.  In your example if each instance needed that array then it will be initialized.  For all practical purposes I think that setting a hard-coded value is sufficient for a couple of reasons.  First, as @michalakes points out even if the size is something very large, the memory footprint will be smaller than a single character array.  Second, adding this to a model side configuration just adds unnecessary complexity.

Reply:
> I respectfully request that it be approved and merged and that that we worry about hypotheticals if and when they arise. Agile.

I understand what you are saying but it would also help if you had followed the published workflow for this project (https://github.com/NCAR/ccpp-framework/wiki/Development-Workflow). Also, the governance committee is supposed to approve changes to the functionality of the framework before they are implemented. I will not be a holdout in the sense if the other reviewers approve, I will as well. However, the issues of non-compliant physics schemes are _not_  hypotheticals. capgen is also not  hypothetical and IMHO anything which makes unification harder should definitely be approved by the CCPP Framework governance committee.

Reply:
@areinecke, 

>  My understanding of the purpose of the logical is to indicate if shared data that gets initialized/ingested only occurs once. In your example if each instance needed that array then it will be initialized.

I think these statements are correct. However, the module variables created by capgen are 2-D or 3-D variables and they are there because the suite modifies them. Every instance will need its own copy of each module variable which can be done but will require quite a bit of work because this is a completely new feature that was not anticipated. Since these are usually 3-D variables, hard coding a large number will indeed be a lot of memory. I do not see how the logical can help with this implementation (but please educate me if I'm still missing something).

Reply:
> > I respectfully request that it be approved and merged and that that we worry about hypotheticals if and when they arise. Agile.
> 
> I understand what you are saying but it would also help if you had followed the published workflow for this project (https://github.com/NCAR/ccpp-framework/wiki/Development-Workflow). Also, the governance committee is supposed to approve changes to the functionality of the framework before they are implemented. I will not be a holdout in the sense if the other reviewers approve, I will as well. However, the issues of non-compliant physics schemes are _not_ hypotheticals. capgen is also not hypothetical and IMHO anything which makes unification harder should definitely be approved by the CCPP Framework governance committee.

You're quite right and I respect your process. Could you (plural) please expedite this through the process. The PR is solid and addresses needed functionality. If there are tweaks needed then please go ahead. We've already done the heavy lifting on this. As to whether the feature was not anticipated, I'd only point out that it supports ensembles and if it wasn't, it should have been.

Reply:
> Understood, however, until now, I have never seen ensembles in the same MPI task so I missed that possibility. I have not been going to any meetings (where I understand this PR has been discussed). Can anyone fill me (or us) in on the current thinking on this PR?

Not sure if this is what you what you are asking (others please chime in), but in addition to supporting ensembles the feature is also needed to support our 4DVAR, from which the tangent linear and non-linear instances of the model are calling the same (but separate instances of) the physics. There are probably other use cases wherein multiple instances of the same physics would need to be invoked within the same process (superparameterizations, maybe?)  Thanks for being open to considering this further.